source/agora/consensus/EnrollmentManager.d

```diff
@@ -1519,7 +1519,7 @@ unittest
     // make the enrollment a validator again
     man.clearExpiredValidators(Height(validator_cycle + 11));
     assert(man.addValidator(enroll, key_pair.address, Height(validator_cycle + 11),
-        &utxo_set.peekUTXO, utxos));
+        &utxo_set.peekUTXO, utxos) == null);    // TODO - check why this was expected to fail
```

Comment:
That is a strange one. I think your change makes sense, but I don't know why it didn't fail before..

Also the other calls to `addValidator` don't seem to check the return value.

Reply:
@omerfirmak  added the assertion so perhaps he can comment on this before this change is accepted.

Reply:
What test is doing right know may not be the original intention But I think that addValidator should fail.

Enrollment is created for `Height(10 + validator_cycle)` but `addValidator` is called with `Height(validator_cycle + 11)`. They should match.

Reply:
In that case I will have a look at my changes to see what broke it. I suspect it is related to the fact that expired validators are not deleted from previous cycle as they are needed for the block signature verification.

Reply:
ValidatorSet in the `v0.x.x` does not delete the most recent enrollment too. I'll take a look at your ValidatorSet changes and let you know if I find something.

Reply:
`findRecentEnrollment` assumed a single entry per key. So that is probably the main source of the issue. https://github.com/bpfkorea/agora/blob/d3c3de35c99d15809fc9510d21c0e531bbd59883/source/agora/consensus/state/ValidatorSet.d#L714

Now that query is not guaranteed to return the most recent enrollment. It can also be the one before the last one. So `findRecentEnrollment` has to check if there is an active enrollment and return that if there is one. If there is no active enrollment it should just return the inactive one which will probably be the only one in the DB :+1:

Reply:
Actually found the issue in clearExpiredValidators where I was adding 1 rather than subtracting 1.

Reply:
I will also check if `findRecentEnrollment` is still working based on your comment.

Reply:
Added order by clause to return active if present.